Fix #6251: Add basic typescript configuration

ankita240796 opened the pull request:
## Explanation
Fixes #6251 

This PR adds a basic typescript configuration and typings for used packages.

ankita240796 committed:
Change way of parsing comments

ankita240796 committed:
Add types install

ankita240796 committed:
Add tsconfig

ankita240796 committed:
Add typescript installation

ankita240796 committed:
Remove comment formatting to avoid duplicate identifier error

ankita240796 committed:
Improve naming

seanlip commented on constants.py:
```diff
@@ -20,14 +20,18 @@
-    first_bracket_index = text.find('= {')
-    last_bracket_index = text.rfind('}')
-    json_text = text[first_bracket_index + 2:last_bracket_index + 1]
+    text_without_comments = re.sub(r'//.*\n', r'\n', text)
```
Optional: Maybe drop the line altogether? I.e. replace `r'\n[two spaces]//.*\n` by `r'\n'`?

(I can't seem to type two spaces in the github comment, it keeps getting collapsed into 1.)

It might also be worth adding some unit tests for this function (or at least the part of it that does the parsing).

ankita240796 replied:
Done (I separated the removing comments part into a new function since a test was already present to check if the json parsed is valid. So, I added a new test which just checks if the new functions for removing comments removes all the comments.)

seanlip commented on tsconfig.json:
```diff
@@ -0,0 +1,7 @@
+  "compilerOptions": {
+    "target": "es5",
+    "typeRoots": ["./node_modules/@types"],
+    "types": ["jasmine", "node", "angular", "jquery", "googlemaps", "q", "selenium-webdriver"]
```
Consider alphabetizing the items in the list.

ankita240796 replied:
Done.

seanlip reviewed (commented):
Hi @ankita240796, thanks! This basically LGTM, couple of small suggestions.

Also, just one general question to double-check my understanding -- does this change the developer workflow at all? If not, when will that happen? I guess it's when the files are changed from .js to .ts (which is when you're planning to add a watcher, right)?

seanlip commented:
Well, maybe not LGTM -- backend tests are failing. (Which does suggest the need for some unit tests for the JSON parsing -- make sure to run those at least before submitting.)

Also, for every change you make (even backend ones), please always test manually prior to submitting a PR (just start up the localhost instance and poke around with it a bit). That will help catch any careless errors.

ankita240796 committed:
Fix failing test

ankita240796 committed:
alphabetize items

ankita240796 committed:
fix space

ankita240796 commented:
Hi @seanlip, 

> does this change the developer workflow at all? If not, when will that happen? I guess it's when the files are changed from .js to .ts (which is when you're planning to add a watcher, right)?

Yes, the change in developer workflow will occur when changing files from .js to .ts

> (Which does suggest the need for some unit tests for the JSON parsing -- make sure to run those at least before submitting.)

Done and sorry for missing on this before.

> Also, for every change you make (even backend ones), please always test manually prior to submitting a PR (just start up the localhost instance and poke around with it a bit). That will help catch any careless errors.

Will be careful about this in future.

I have addressed the review comments, PTAL. Thanks!

seanlip commented on constants_test.py:
```diff
@@ -44,3 +45,10 @@ def test_constants_and_feconf_are_consistent(self):
+
+    def test_all_comments_are_removed_from_json_text(self):
+        """Tests if comments are removed from json text."""
+        with open(os.path.join('assets', 'constants.js'), 'r') as f:
```
Don't do this -- the file will change, and because of that, the test is quite weak.

Make a dummy simple test file in core/tests/data instead. Then assert that its contents after the operation are what you expect (compare it to a "golden string").

ankita240796 replied:
Done.

ankita240796 committed:
Add dummy constants for testing comment removal

ankita240796 commented:
Hi @seanlip, I have created a dummy file for testing comment removal, PTAL. Thanks!

seanlip reviewed (approved):
LGTM. Thanks!
